File: frontend/src/types/portfolio.ts
// TypeScript types for the Portfolio Generator API

export interface PortfolioSkill {
  skill_name: string;
  category: string;
  confidence_score: number;
}

export interface SkillGroup {
  category: string;
  skills: PortfolioSkill[];
}

export interface PortfolioProject {
  id: number;
  name: string;
  description: string | null;
  repo_url: string;
  language: string | null;
  stars: number;
  forks: number;
  // Scores
  build_score: number | null;
  proof_score: number | null;
  complexity_score: number | null;
  security_score: number | null;
  documentation_score: number | null;
  // Contribution
  contribution_percentage: number | null;
  contribution_summary: string | null;
  primary_contributor: string | null;
  // Deployment
  deployment_url: string | null;
  deployment_reachable: boolean | null;
  deployment_score: number | null;
  provider: string | null;
  // Insights
  project_type: string | null;
  complexity_level: string | null;
  project_summary: string | null;
  technical_summary: string | null;
  project_categories: string[];
  // Skills
  skills: PortfolioSkill[];
  frameworks: string[];
}

export interface PortfolioHeader {
  username: string;
  name: string | null;
  avatar_url: string | null;
  github_url: string;
  repositories_verified: number;
  avg_build_score: number;
  avg_proof_score: number;
}

export interface VerificationSummary {
  repositories_analyzed: number;
  deployments_verified: number;
  verified_skills_count: number;
  primary_contributor_projects: number;
}

export interface PortfolioResponse {
  header: PortfolioHeader;
  verification_summary: VerificationSummary;
  skill_groups: SkillGroup[];
  projects: PortfolioProject[];
}

File: backend/app/schemas/portfolio.py
"""
Pydantic schemas for the Portfolio Generator API.
"""
from __future__ import annotations
from typing import List, Optional
from pydantic import BaseModel


class PortfolioSkill(BaseModel):
    skill_name: str
    category: str
    confidence_score: int

    model_config = {"from_attributes": True}


class SkillGroup(BaseModel):
    category: str
    skills: List[PortfolioSkill]


class PortfolioDeployment(BaseModel):
    deployment_url: str
    provider: Optional[str] = None
    reachable: bool
    deployment_score: int
    ssl_enabled: bool

    model_config = {"from_attributes": True}


class PortfolioProject(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    repo_url: str
    language: Optional[str] = None
    stars: int
    forks: int
    # Scores
    build_score: Optional[int] = None
    proof_score: Optional[int] = None
    complexity_score: Optional[int] = None
    security_score: Optional[int] = None
    documentation_score: Optional[int] = None
    # Contribution
    contribution_percentage: Optional[float] = None
    contribution_summary: Optional[str] = None
    primary_contributor: Optional[str] = None
    # Deployment
    deployment_url: Optional[str] = None
    deployment_reachable: Optional[bool] = None
    deployment_score: Optional[int] = None
    provider: Optional[str] = None
    # Insights
    project_type: Optional[str] = None
    complexity_level: Optional[str] = None
    project_summary: Optional[str] = None
    technical_summary: Optional[str] = None
    project_categories: List[str] = []
    # Skills
    skills: List[PortfolioSkill] = []
    frameworks: List[str] = []


class PortfolioHeader(BaseModel):
    username: str
    name: Optional[str] = None
    avatar_url: Optional[str] = None
    github_url: str
    repositories_verified: int
    avg_build_score: float
    avg_proof_score: float


class VerificationSummary(BaseModel):
    repositories_analyzed: int
    deployments_verified: int
    verified_skills_count: int
    primary_contributor_projects: int


class PortfolioResponse(BaseModel):
    header: PortfolioHeader
    verification_summary: VerificationSummary
    skill_groups: List[SkillGroup]
    projects: List[PortfolioProject]
